File: 2670/go/main.go
func distinctDifferenceArray(nums []int) []int {
	n := len(nums)
	result := make([]int, n)
	suffixCountArray := make([]int, n+1)
	s := map[int]int{}
	for i := n - 1; i > 0; i-- {
		s[nums[i]] = int(1)
		suffixCountArray[i] = len(s)
	}
	s = map[int]int{}
	for i := 0; i < n; i++ {
		s[nums[i]] = int(1)
		result[i] = len(s) - suffixCountArray[i+1]
	}
	return result
}

File: 2670/python3/main.py
class Solution:
    def distinctDifferenceArray(self, nums: list[int]) -> list[int]:
        n = len(nums)
        s = set()
        suffixCountArray = [0] * (n + 1)
        for i in range(n - 1, 0, -1):
            s.add(nums[i])
            suffixCountArray[i] = len(s)
        res = [0] * n
        s = set()
        for i in range(n):
            s.add(nums[i])
            res[i] = len(s) - suffixCountArray[i + 1]
        return res
